src/lib/core/common-behaviors/id.ts

```diff
@@ -0,0 +1,35 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Constructor} from './constructor';
+
+/** Monotonically increasing integer used to auto-generate unique ids. */
+let nextUniqueId = 0;
+
+/** @docs-private */
+export interface CanId {
+  id: string;
+}
+
+/**
+ * Mixin to augment a directive with an `id` property.
+ * The id property will fall back to auto-generated unique ids if developers don't specify one.
+ */
+export function mixinId<T extends Constructor<{}>>(base: T, componentName: string)
```

Comment:
Hmm I see what you mean. I just had the impression (due to #5394) that it would be helpful to fall back to the auto-generated id if it's set to a falsy value. 

I'm fine changing it to throw an error.